File: leetcode_py/200.py
from typing import List, Optional
import sys

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        result = 0
        for i, row in enumerate(grid):
            for j, ele in enumerate(row):
                if ele == "1":
                    self.dfs(grid, i, j)
                    result += 1
        return result
    def dfs(self, grid, i, j):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] != "1":
            return
        grid[i][j] = '2'
        self.dfs(grid, i - 1, j)
        self.dfs(grid, i + 1, j)
        self.dfs(grid, i, j - 1)
        self.dfs(grid, i, j + 1)

if __name__ == '__main__':
    grid = [["1","1","1","1","0"],
            ["1","1","0","1","0"],
            ["1","1","0","0","0"],
            ["0","0","0","0","0"]]
    sol = Solution()
    print(sol.numIslands(grid))

File: leetcode_cpp/200.cpp
/*
岛屿数量 图论/DFS
给你一个由 '1'（陆地）和 '0'（水）组成的的二维网格，请你计算网格中岛屿的数量。
岛屿总是被水包围，并且每座岛屿只能由水平方向和/或竖直方向上相邻的陆地连接形成。
此外，你可以假设该网格的四条边均被水包围。
*/
using namespace std;
#include <vector>
#include <iostream>
#include <algorithm>
#include <numeric>
#include <unordered_map>
#include <cmath>
#include <queue>
#include <stack>
#include <functional>

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int result = 0;
        auto is_out = [&](int r, int c)->bool {
            return !(r >= 0 && r < grid.size() && c >= 0 && c < grid[0].size() && grid[r][c] == '1');
        };
        function<void(int, int)> dfs = [&](int r, int c)->void {
            if(is_out(r, c)) {
                return;
            }
            grid[r][c] = '2';
            dfs(r - 1, c);
            dfs(r + 1, c);
            dfs(r, c - 1);
            dfs(r, c + 1);
        };
        for(int i = 0; i < grid.size(); ++i) {
            for(int j = 0; j < grid[i].size(); ++j) {
                if(grid[i][j] == '1') {
                    dfs(i, j);
                    result++;
                }
            }
        }
        return result;
    }
};

int main()
{
    vector<vector<char>> grid = {{'1','1','0','0','0'}, {'1','1','0','0','0'}, {'0','0','1','0','0'}, {'0','0','0','1','1'}};
	Solution sol;
	int out = sol.numIslands(grid);
	cout << out << endl;
	system("pause");
}


